### ISSUE TITLE
v5 Request - Could you think about dropping the custom delegate types and just use normal EventHandler with models?

### ISSUE THREAD

**Issue by grofit** [BarRaider/obs-websocket-dotnet#99]
**Issue Type** 
- Feature Request

**Describe the issue**
Currently most of the events that you can hook into are custom delegates which at face value may seem useful but in real use cases it makes it trickier to do any sort of event proxying approaches. It also makes data more fragmented as there isnt a singular model representing the response data.

If normal `EventHandler<SomeResponseModel>` was used it simplifies the events and makes it simpler to just have a singular payload which contains the data you care about.

To give an example of why this is a problem, if you were to use RX to add things like event throttling or delays etc you cannot simply wrap them without doing a lot of legwork wrapping all the custom delegates and args, and if its has more than 1 response which contains complex models its a massive pain binding all that.

```csharp
// Example binding to a custom delegate event i.e 
//  public delegate void CustomDelegate(A a, B b);
Observable
   .FromEvent<CustomDelegate, CustomOutput>(  // Need to handle both delegate and event/args, also need to make my own CustomOutput model
       h => (a, b) => h(new CustomOutput{ A= a, B = b }), // If there isnt a singular response arg its even more painful
       h => someObject.someEvent += h,
       h => someObject.someEvent -= h)
    .Select(x => x.EventArgs); // Outputs Observable<CustomOutput>

// Example binding to a EventHandler<CustomOutput> event
Observable
   .FromEventPattern<CustomOutput>( // Using normal EventHandler no need to faff
        h=> someObject.someEvent += h,
        h => someObject.someEvent -= h)
    .Select(x => x.EventArgs); // Outputs Observable<CustomOutput>
```
**Versions**
OBS Version: N/A
OBS WebSocket Version: 5.x
OBS WebSocket Dotnet (this library) Version: 5.x
OS: N/A

**Additional context**
There was a PR which was done against the v4 which did this and many other things such as Async method handling but unfortunately due to the PR being one big thing vs lots of smaller chunks it was discarded, but for v4 I have had to use the fork of this library because it took me 10 minutes to wrap all existing events into observables when they were `EventHandler<T>` delegates, and I lost 2 hours trying to make all the required classes/wrapper logic for about 5 events with the custom delegates and gave up.

**Comment by BarRaider** [comment]
Overall I don't have strong objections against this, however don't have the bandwidth for this.

**Comment by grofit** [comment]
Same as other comment, as this will hopefully be a simple-ish change and @Zingabopp has already done the work in v4, if you are open to it I will look at raising a PR for it when I get time.

**Comment by grofit** [comment]
I have started work on this one, a few things to mention:

- I am making everything its own event model even if some could be shared, i.e transition started/ended, I was thinking about sharing some event models but going forward if an event changes you would have to create separate ones anyway so for the extra couple of classes it at least makes it easier to change them going forward if needed.

- Im using classes vs records as its still netstandard2.0 so C# 9 cant be used unless a higher target framework is used.

- I am trying to keep the comments that you already have wherever possible just transposing them over

- I am putting all event models into `Types/Events` namespace, I was going to put it root level but as you have a `Types` namespace already it made sense to put them there, but easy enough to change if you have feelings either way

I will hopefully have a PR ready later today for you to look over, its worth noting this will be a breaking change as anyone using events would now have to adhere to the new sig i.e:

```csharp
// OLD
private void OBS_onSceneTransitionVideoEnded(OBSWebsocket sender, string transitionName)
{
    LogMessage($"[SceneTransitionVideoEnded] Name: {transitionName}");
}

// NEW
private void OBS_onSceneTransitionVideoEnded(object sender, SceneTransitionVideoEndedEventArgs args)
{
    LogMessage($"[SceneTransitionVideoEnded] Name: {args.TransitionName}");
}
```

Here is an example of a larger event object keeping comments as close as possible to your delegate ones:

```csharp
namespace OBSWebsocketDotNet.Types.Events
{
    /// <summary>
    /// Event args for <see cref="OBSWebsocket.SceneItemCreated"/>
    /// </summary>
    public class SceneItemCreatedEventArgs
    {
        /// <summary>
        /// Name of the scene where the item is
        /// </summary>
        public string SceneName { get; } 
        
        /// <summary>
        /// Name of the concerned item
        /// </summary>
        public string SourceName { get; }
        
        /// <summary>
        /// Numeric ID of the scene item
        /// </summary>
        public int SceneItemId { get; } 
        
        /// <summary>
        /// Index position of the item
        /// </summary>
        public int SceneItemIndex { get; }

        public SceneItemCreatedEventArgs(string sceneName, string sourceName, int sceneItemId, int sceneItemIndex)
        {
            SceneName = sceneName;
            SourceName = sourceName;
            SceneItemId = sceneItemId;
            SceneItemIndex = sceneItemIndex;
        }
    }
}
```

If you have any concerns so far let me know

**Comment by BarRaider** [comment]
Overall looks good and aligns with what we discussed.

**Comment by grofit** [comment]
I have finished this work and raised it in #106 let me know if you want any changes, I am in a bit of a rush for this as it unblocks some other work im doing, but I appreciate you will need time to look it over etc, so quicker I can make remediations the better.

There is one thing I may recommend as a subsequent change which is change all the `List<T>` properties in the event args over to `IReadOnlyCollection<T>` just to further enforce that these event args are immutable.

**Comment by BarRaider** [comment]
Fixed in #106
